handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,492 @@
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.local.LocalAddress;
+import io.netty.channel.local.LocalChannel;
+import io.netty.channel.local.LocalServerChannel;
+import io.netty.handler.ssl.OpenSsl;
+import io.netty.handler.ssl.ReferenceCountedOpenSslEngine;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.SslProvider;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import io.netty.util.ReferenceCountUtil;
+
+import java.net.SocketAddress;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.net.ssl.SSLHandshakeException;
+
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNotSame;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assume.assumeTrue;
+
+public class OcspTest {
+
+    @BeforeClass
+    public static void checkOcspSupported() {
+        assumeTrue(OpenSsl.isOcspSupported());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testJdkClientEnableOcsp() throws Exception {
+        SslContextBuilder.forClient()
+                .sslProvider(SslProvider.JDK)
+                .enableOcsp(true)
+                .build();
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testJdkServerEnableOcsp() throws Exception {
+        SelfSignedCertificate ssc = new SelfSignedCertificate();
+        try {
+            SslContextBuilder.forServer(ssc.certificate(), ssc.privateKey())
+                    .sslProvider(SslProvider.JDK)
+                    .enableOcsp(true)
+                    .build();
+        } finally {
+            ssc.delete();
+        }
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testClientOcspNotEnabledOpenSsl() throws Exception {
+        testClientOcspNotEnabled(SslProvider.OPENSSL);
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testClientOcspNotEnabledOpenSslRefCnt() throws Exception {
+        testClientOcspNotEnabled(SslProvider.OPENSSL_REFCNT);
+    }
+
+    private void testClientOcspNotEnabled(SslProvider sslProvider) throws Exception {
+        SslContext context = SslContextBuilder.forClient()
+                .sslProvider(sslProvider)
+                .build();
+        try {
+            SslHandler sslHandler = context.newHandler(ByteBufAllocator.DEFAULT);
+            ReferenceCountedOpenSslEngine engine = (ReferenceCountedOpenSslEngine) sslHandler.engine();
+            engine.getOcspResponse();
```

Comment:
the `engine` must be released at some point. `SslHandler` typically takes ownership, but it must be interred into a pipeline ignorer to do so.